### PR TITLE
Tor with Snowflakes as experimental feature

### PR DESCRIPTION
**Pull request by karliatto**
## Description

In some countries ISPs have started blocking Trezor servers, in the same countries  using Tor directly is not possible, it is blocked at the ISP level as well. 

Tor snowflakes (https://snowflake.torproject.org/) is "pluggable" transport of Tor that helps user to circumvent  those prohibitions. 

* Make Tor run with snowflakes as a proxy integrated in Suite
* Repository with all the binaries https://www.torproject.org/download/tor/
* There could be a feature that will download the binary for the architecture and use it.

image

Proper documentation about what is required to run it will be added to Trezor Knowledge Base but if you are reviewing it now you can follow steps bellow to get the binary.

```
* Use it with Debian
      - Run "apt install snowflake-client"
      - Then use the path to the binary "/usr/bin/snowflake-client"

      or:
      - Download it from https://packages.debian.org/bookworm/amd64/snowflake-client/download
      - Install it like sudo dpkg -i snowflake-client_2.5.1-1+b3_amd64.deb
      - Then use the path to the binary "/usr/bin/snowflake-client"

* Use it with macOS:
      - Run "brew install snowflake"
      - Then use the path to the binary "/opt/homebrew/bin/snowflake-client"

* Use it with Windows:

      - Download Tor Browser from [here](https://www.torproject.org/dist/torbrowser/13.5.1/tor-browser-windows-x86_64-portable-13.5.1.exe).
      - Install it. By default, it installs (extracts) on the Desktop.
      - The binary is located in "C:\Users\USERNAME\Desktop\Tor Browser\Browser\TorBrowser\Tor\PluggableTransports"
```

### CODE REVIEW

**Review by mroz22 — COMMENTED** [review]
are the client files automatically checked out to git-lfs? 
I guess they are but I am not very skilled with git-lfs so maybe i am doing something wrong but I can't see it there 
```git lfs ls-files --all | grep snowflake```

**Comment by karliatto** [comment]
> are the client files automatically checked out to git-lfs? I guess they are but I am not very skilled with git-lfs so maybe i am doing something wrong but I can't see it there `git lfs ls-files --all | grep snowflake`

@mroz22 you are right it was not in lfs but I just added it. Thanks!

File: packages/request-manager/src/controller.ts [hunk 1]
```diff
@@ -66,11 +66,17 @@ export class TorController extends EventEmitter {
...
-    public getTorConfiguration(processId: number): string[] {
-        const controlAuthCookiePath = path.join(this.options.torDataDir, 'control_auth_cookie');
+    public getTorConfiguration(
+        processId: number,
+        useSnowflake?: boolean,
+        snowflakeBinaryPath?: string,
+    ): string[] {
+        const { torDataDir } = this.options;
+        const controlAuthCookiePath = path.join(torDataDir, 'control_auth_cookie');
+        const snowflakeLogPath = path.join(this.options.torDataDir, 'snowflake.log');
```
Comment (marekrjpolak) [inline packages/request-manager/src/controller.ts]: nit:
```suggestion
        const snowflakeLogPath = path.join(torDataDir, 'snowflake.log');
```

File: packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import { Input, Switch } from '@trezor/components';
```
Comment (marekrjpolak) [inline packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx]: nit: typo in file name

File: packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            }
+        });
+
+        desktopApi.getTorSettings().then(result => {
+            if (result.success) {
+                setTorSettings(result.payload);
+            } else {
+                setError(result.error);
+            }
+        });
```
Comment (marekrjpolak) [inline packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx]: Rebase issue?

File: packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                                shouldUseSnowflake: !torSettings.shouldUseSnowflake,
+                            })
+                        }
+                    />
+                </ActionColumn>
+            </SectionItem>
+            <SectionItem data-test="@settings/general/tor/snowflake-binary-path">
+                <TextColumn
+                    title="Path for Tor Snowflake"
+                    description="This should be the path of Snowflake in your system."
```
Comment (marekrjpolak) [inline packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx]: Missing translations, but it's probably kinda ok for experimental features.

**Review by marekrjpolak — COMMENTED** [review]
I'm not quite sure about separate `shouldUseSnowflake` bool and **required** `snowflakeBinaryPath` string.

First, I'm not sure whether it makes sense to have binary path set but the whole snowflake turned off, or to have it turned on without valid binary path.

Second, what happened to me: I already had `torSettings` in main (=> no fallback to default). When I opened Suite settings, I received these stored settings, without the flag and binary path. I tried to turn snowflake on (which could work even with `shouldUseSnowflake` undefined), but the message was blocked by `suite-desktop-api` because it's invalid without `snowflakeBinaryPath`, which was never set anywhere yet.

Wouldn't it make sense to have it as an input with save button (and therefore also not sending every typed character to desktop-api, for example as in Device/Customization/Device name) and either have a string there (= snowflake on) or not (snowflake off)? And if not, at least make the `snowflakeBinaryPath` optional, instead of having `''` as a default value?

(Note: I still haven't tried whether it works or not as a whole, these were just first observations of mine)

File: packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx (hunk 3 above)
Comment (karliatto) [reply packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx]: Yes, it should be fixed now https://github.com/trezor/trezor-suite/pull/12991/commits/35d42e1dda383e2601304466c42ea2d0b9bb7c76

File: packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx (hunk 4 above)
Comment (karliatto) [reply packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx]: This text is not used anymore in the new version https://github.com/trezor/trezor-suite/pull/12991/commits/35d42e1dda383e2601304466c42ea2d0b9bb7c76#diff-3227a1b9781609288bd6371e791b3b03d4f459f1f231763b6891cb774da430e6

File: packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx (hunk 2 above)
Comment (karliatto) [reply packages/suite/src/views/settings/SettingsGeneral/TorSnoflake.tsx]: Thanks! Fixed https://github.com/trezor/trezor-suite/pull/12991/commits/35d42e1dda383e2601304466c42ea2d0b9bb7c76#diff-3227a1b9781609288bd6371e791b3b03d4f459f1f231763b6891cb774da430e6

File: packages/request-manager/src/controller.ts (hunk 1 above)
Comment (karliatto) [reply packages/request-manager/src/controller.ts]: Right! dd4ab4d899c527a21728300159673622a222b18d

**Comment by karliatto** [comment]
> I'm not quite sure about separate `shouldUseSnowflake` bool and **required** `snowflakeBinaryPath` string.
> 
> First, I'm not sure whether it makes sense to have binary path set but the whole snowflake turned off, or to have it turned on without valid binary path.
> 
> Second, what happened to me: I already had `torSettings` in main (=> no fallback to default). When I opened Suite settings, I received these stored settings, without the flag and binary path. I tried to turn snowflake on (which could work even with `shouldUseSnowflake` undefined), but the message was blocked by `suite-desktop-api` because it's invalid without `snowflakeBinaryPath`, which was never set anywhere yet.
> 
> Wouldn't it make sense to have it as an input with save button (and therefore also not sending every typed character to desktop-api, for example as in Device/Customization/Device name) and either have a string there (= snowflake on) or not (snowflake off)? And if not, at least make the `snowflakeBinaryPath` optional, instead of having `''` as a default value?
> 
> (Note: I still haven't tried whether it works or not as a whole, these were just first observations of mine)

I made it in "2 steps", path and enable, because putting myself in the foot of the final user I thought it was easier to understand.  I am still not sure so if you really think it will be better with only path and when path is there then it means it has to use snowflake let me know.

But I made couple of changes:
1. Follow your suggestion and implemented it similar to Device name so it does send it to desktop-api when enable
2. Add a check that it has to be a "full path"
3. Add a check in request-manager that checks if the file exists and it is actually a file otherwise it defaults to normal Tor without it.

File: packages/utils/src/isFullPath.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+export const isFullPath = (path: string) => {
+    // Full Unix-like and Windows paths
+    const fullPathPattern = /^(\/|([a-zA-Z]:\\))/;
+
+    // Check to avoid matching URLs
+    const urlPattern = /^(https?:\/\/|www\.)/;
```
Comment (marekrjpolak) [inline packages/utils/src/isFullPath.ts]: It seems that the intersection of these two regexes is empty, therefore the second one is not needed, is it possible?

File: packages/node-utils/src/tests/checkFileExists.test.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import { checkFileExists } from '../checkFileExists';
```
Comment (marekrjpolak) [inline packages/node-utils/src/tests/checkFileExists.test.ts]: Eslint warnings:
``` There should be at least one empty line between import groups ```
``` `../checkFileExists` import should occur after import of `path` ```

File: packages/request-manager/e2e/interceptor.test.ts [hunk 7]
```diff
@@ -44,8 +45,9 @@ describe('Interceptor', () => {
             port,
             controlPort,
             torDataDir,
+            snowflakeBinaryPath,
         });
-        const torParams = torController.getTorConfiguration(processId);
+        const torParams = await torController.getTorConfiguration(processId, false);
```
Comment (marekrjpolak) [inline packages/request-manager/e2e/interceptor.test.ts]: The second param probably shouldn't be bool anymore.

File: packages/request-manager/src/controller.ts [hunk 8]
```diff
@@ -2,6 +2,7 @@ import { EventEmitter } from 'events';
 import path from 'path';
 
 import { createTimeoutPromise } from '@trezor/utils';
+import { checkFileExists } from '@trezor/node-utils';
```
Comment (marekrjpolak) [inline packages/request-manager/src/controller.ts:5]: Eslint error:
```'@trezor/node-utils' should be listed in the project's dependencies. Run 'npm i -S @trezor/node-utils' to add it```

File: packages/utils/src/isFullPath.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+
+    // Check for full path patterns, excluding URLs
+    const isFullPath = fullPathPattern.test(path);
```
Comment (marekrjpolak) [inline packages/utils/src/isFullPath.ts]: Eslint error:
```'isFullPath' is already declared in the upper scope on line 1 column 14.```

File: packages/suite-desktop-core/src/libs/processes/TorProcess.ts [hunk 10]
```diff
@@ -1,13 +1,8 @@
 import { TorController, TOR_CONTROLLER_STATUS } from '@trezor/request-manager';
 
-import { BaseProcess, Status } from './BaseProcess';
+import { TorConnectionOptions } from '@trezor/request-manager/src/types';
```
Comment (marekrjpolak) [inline packages/suite-desktop-core/src/libs/processes/TorProcess.ts:2]: Eslint warning:
```There should be no empty line within import group```

**Review by marekrjpolak — COMMENTED** [review]
I'm not sure whether it's issue of my setup, Tor network or something else, but when I set the correct snowflake binary path, Tor never connects and I'm not able to turn it off until Suite restart.

Also, I don't know how to turn the snowflake off now. 
image

File: packages/request-manager/src/controller.ts (hunk 8 above)
Comment (karliatto) [reply packages/request-manager/src/controller.ts:5]: https://github.com/trezor/trezor-suite/pull/12991/commits/c53bdc9fcf78fcf83c266fad4f8d5792eb01e73e

File: packages/request-manager/e2e/interceptor.test.ts (hunk 7 above)
Comment (karliatto) [reply packages/request-manager/e2e/interceptor.test.ts]: https://github.com/trezor/trezor-suite/pull/12991/commits/c53bdc9fcf78fcf83c266fad4f8d5792eb01e73e

File: packages/node-utils/src/tests/checkFileExists.test.ts (hunk 6 above)
Comment (karliatto) [reply packages/node-utils/src/tests/checkFileExists.test.ts]: 32c6d71fcced0dd60bedb4477e70344ba211d51a

File: packages/suite-desktop-core/src/libs/processes/TorProcess.ts (hunk 10 above)
Comment (karliatto) [reply packages/suite-desktop-core/src/libs/processes/TorProcess.ts:2]: 449369ec9a3e6ac35f219dfd1f76a3833003e5c1

File: packages/utils/src/isFullPath.ts (hunk 5 above)
Comment (karliatto) [reply packages/utils/src/isFullPath.ts]: What about a simplification like https://github.com/trezor/trezor-suite/pull/12991/commits/6a40e46a6c52c8df049c4d8d45434be4adc8651a  ?

File: packages/utils/src/isFullPath.ts (hunk 9 above)
Comment (karliatto) [reply packages/utils/src/isFullPath.ts]: Let's simplify it like https://github.com/trezor/trezor-suite/commit/6a40e46a6c52c8df049c4d8d45434be4adc8651a

**Comment by karliatto** [comment]
> Also, I don't know how to turn the snowflake off now.

Good point, what about when changing from some path to empty string it will display "Disable Snowflake" in the button to save it like there 589e43f8dd8cc5e0237accc96bd35b84815e02b5  ?

image

File: packages/utils/src/isFullPath.ts (hunk 5 above)
Comment (marekrjpolak) [reply packages/utils/src/isFullPath.ts]: Excellent!

**Comment by Ondra-Zik** [comment]
> Good poin

Also, maybe I'm commenting a bit late, so it's okay if it is not possible, but could there be a file picker so you are not forced to copy the path from file explorer?

**Comment by marekrjpolak** [comment]
> could there be a file picker

Now I'm just genuinely amazed that it didn't cross my mind until you mentioned it @Ondra-Zik-SL :sweat_smile:

**Review by marekrjpolak — APPROVED** [review]
I can confirm that it works and I believe it's good enough for experimental features. Good job!

I agree with @Ondra-Zik-SL that the file picker is nice-to-have, but it's definitely not necessary for my approval.

**Comment by karliatto** [comment]
> > Good poin
> 
> Also, maybe I'm commenting a bit late, so it's okay if it is not possible, but could there be a file picker so you are not forced to copy the path from file explorer?

That is a good point, I didn't think about it neither. :sweat_smile: I will merge with current approach and will iterate over the file picker in a follow-up.

Thanks both @marekrjpolak and @Ondra-Zik-SL for the feedback!

**Comment by karliatto** [comment]
@Ondra-Zik-SL it is actually not possible to use the `<input type="file">` from HTML due to browser security features it does not allow you to get the full path, just the file object that can be used for uploading for example, but not for our purposes. Instead the browser replaces the path with default "C:\fakepath\" for historical reasons. More details in link below:

https://html.spec.whatwg.org/multipage/input.html#fakepath-srsly

**Pull request merged**